cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+    /// repository option as well. This is typically a full SSH URL for your
+    /// Gerrit instance.
+    #[arg(long)]
+    remote: Option<String>,
+
+    /// If true, do not actually push the changes to Gerrit.
+    #[arg(long = "dry-run", short = 'n')]
+    dry_run: bool,
+}
+
+/// calculate push remote. The logic is:
+/// 1. If the user specifies `--remote`, use that
+/// 2. If the user has 'gerrit.remote' configured, use that
+/// 3. If there is a default push remote, use that
+/// 4. If the user has a remote named 'gerrit', use that
+/// 5. otherwise, bail out
+fn calculate_push_remote(
+    store: &Arc<Store>,
+    config: &UserSettings,
+    remote: Option<String>,
+) -> Result<String, CommandError> {
+    let git_repo = git::get_git_repo(store)?; // will fail if not a git repo
+    let remotes = git_repo.remote_names();
+
+    // case 1
+    if let Some(remote) = remote {
+        if remotes.contains(BStr::new(&remote)) {
+            return Ok(remote);
+        }
+        return Err(user_error(format!(
+            "The remote '{remote}' (specified via `--remote`) does not exist",
+        )));
+    }
+
+    // case 2
+    if let Ok(remote) = config.get_string("gerrit.default-remote") {
+        if remotes.contains(BStr::new(&remote)) {
+            return Ok(remote);
+        }
+        return Err(user_error(format!(
+            "The remote '{remote}' (configured via `gerrit.default-remote`) does not exist",
+        )));
+    }
+
+    // case 3
+    if let Some(remote) = git_repo.remote_default_name(gix::remote::Direction::Push) {
+        return Ok(remote.to_string());
+    }
+
+    // case 4
+    if remotes.iter().any(|r| **r == "gerrit") {
+        return Ok("gerrit".to_owned());
+    }
+
+    // case 5
+    Err(user_error(
+        "No remote specified, and no 'gerrit' remote was found",
+    ))
+}
+
+/// Determine what Gerrit ref and remote to use. The logic is:
+///
+/// 1. If the user specifies `--remote-branch branch`, use that
+/// 2. If the user has 'gerrit.default-remote-branch' configured, use that
+/// 3. Otherwise, bail out
+fn calculate_push_ref(
+    config: &UserSettings,
+    remote_branch: Option<String>,
+) -> Result<String, CommandError> {
+    // case 1
+    if let Some(remote_branch) = remote_branch {
+        return Ok(remote_branch);
+    }
+
+    // case 2
+    if let Ok(branch) = config.get_string("gerrit.default-remote-branch") {
+        return Ok(branch);
+    }
+
+    // case 3
+    Err(user_error(
+        "No target branch specified via --remote-branch, and no 'gerrit.default-remote-branch' \
+         was found",
+    ))
+}
+
+pub fn cmd_upload(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    upload: &UploadArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    let revisions: Vec<_> = workspace_command
+        .parse_union_revsets(ui, &upload.revisions)?
+        .evaluate_to_commits()?
```

Comment:
Can we collect to `evaluate_to_commit_ids()` instead? I didn't see anywhere we need the commits, but maybe I just missed it.